### PR TITLE
Calculate num_frames if missing from metadata

### PR DESCRIPTION
**Pull request by Dan-Flores**
This PR enables the instantiation of a VideoDecoder without `num_frames_from_content` or `num_frames_from_header` in the metadata, to resolve issue #727. Without these variables, the VideoDecoder will calculate the number of frames using `average_fps_from_header` and `duration_seconds_from_header`.

### Tests in test_decoders.py
* The return value of `_get_stream_json_metadata` is mocked to ensure VideoDecoder is instantiated without `num_frames_from_content` or `num_frames_from_header`. 
* Various `get_frame...` methods are tested on this decoder.

### Tests in test_metadata.py
* `test_calculate_num_frames_using_fps_and_duration` instantiates `VideoStreamMetadata` without `num_frames_from_content` or `num_frames_from_header`, and  checks that various values of `average_fps` and `duration_seconds` result in the correct number of frames.

* In `test_num_frames_fallback`, the test where both `num_frames_from_content` and `num_frames_from_header` are `None` is no longer valid. A similar test case is added in `test_calculate_num_frames_using_fps_and_duration`.

### CODE REVIEW

File: src/torchcodec/_core/_metadata.py [hunk 1]
```diff
@@ -129,8 +129,15 @@ def num_frames(self) -> Optional[int]:
         """
```
Comment (scotts) [inline src/torchcodec/_core/_metadata.py:140]: In the doc string, we should also say what we do if num_frames is missing from all of the metadata.

File: src/torchcodec/_core/_metadata.py [hunk 2]
```diff
@@ -129,8 +129,15 @@ def num_frames(self) -> Optional[int]:
...
         if self.num_frames_from_content is not None:
             return self.num_frames_from_content
-        else:
+        elif self.num_frames_from_header is not None:
             return self.num_frames_from_header
+        elif (
+            self.average_fps_from_header is not None
+            and self.duration_seconds_from_header is not None
+        ):
```
Comment (scotts) [inline src/torchcodec/_core/_metadata.py:148]: I just went down an interesting rabbit-hole of asking "Should we instead use the `average_fps` property?" And the answer is no! :) If `self.average_fps_from_header is None`, then `self.average_fps` actually calls this property, `self.num_frames` and we'd (I think) hit infinite recursion. I do wonder if there is a way for us to be more systematic about this, where we could more centrally define the logic of under what circumstances we use which metadata.

File: test/test_decoders.py [hunk 3]
```diff
@@ -738,6 +740,76 @@ def test_get_frames_in_range(self, stream_index, device, seek_mode):
...
+        frames = decoder.get_frames_at(indices)
+        for ref, frame in zip(ref_frames, frames.data):
+            torch.testing.assert_close(ref, frame)
+
+        # Test get_frames_played_in_range to get all frames
+        assert decoder.metadata.end_stream_seconds is not None
+        all_frames = decoder.get_frames_played_in_range(
+            decoder.metadata.begin_stream_seconds, decoder.metadata.end_stream_seconds
+        )
+        assert_frames_equal(all_frames.data, decoder[:])
```
Comment (scotts) [inline test/test_decoders.py]: Can we also add a simple `assert len(decoder) == CORRECT_VALUE`?

**Review by scotts — CHANGES_REQUESTED** [review]
I had initially approved, but it looks like the tests are failing on Mac - let's make sure we're using the right tolerances there.

**Comment by scotts** [comment]
Thank you! Also, this is only half of #727. We also want to be able to handle if `num_frames` is present but `duration` is not. This half, though, is definitely the more valuable side, as core operations in the Python `VideoDecoder` depend on `num_frames`.

File: test/test_decoders.py [hunk 4]
```diff
@@ -738,6 +740,77 @@ def test_get_frames_in_range(self, stream_index, device, seek_mode):
...
+        assert decoder.metadata.num_frames_from_header is None
+        assert decoder.metadata.num_frames_from_content is None
+        assert decoder.metadata.duration_seconds is not None
+        assert decoder.metadata.average_fps is not None
+        assert decoder.metadata.num_frames == int(
+            decoder.metadata.duration_seconds * decoder.metadata.average_fps
+        )
+        assert len(decoder) == 390
+
+        # Test get_frames_in_range
```
Comment (scotts) [inline test/test_decoders.py]: Let's add a comment here that we're really just testing the logic in the Python `VideoDecoder` class when we do the decoding. We do checks that involve `num_frames` at the Python layer, and that logic will use the mocked metadata. But the C++ layer does not use that mocked metadata, and will instead use what it read from the video file itself.

We should also create an issue about adding a video file that has the number of frames missing in its metadata so we can actually test the C++ logic as well. This test would then use that video file, and not need any mocking.

**Comment by scotts** [comment]
Thanks, @Dan-Flores! I looked into the failures, and they're all known (#740, #734, #724).

File: test/test_decoders.py (hunk 4 above)
Comment (Dan-Flores) [reply test/test_decoders.py]: Good point! I'll remove some unnecessary python calls here, and add a comment indicating that the python logic is the only aspect being tested.

It looks like issue #710 covers this subject.

**Comment by Dan-Flores** [comment]
I've updated this PR to complete the other half of #727, and updated the description as well.

File: test/test_metadata.py [hunk 5]
```diff
@@ -143,6 +143,93 @@ def test_num_frames_fallback(
...
+        average_fps_from_header=average_fps_from_header,
+        stream_index=0,
+    )
+
+    assert metadata.num_frames == expected_num_frames
+
+
+@pytest.mark.parametrize(
+    "duration_seconds_from_header, begin_stream_seconds_from_content, end_stream_seconds_from_content, expected_duration_seconds",
+    [(60, 5, 20, 15), (60, 1, None, 60), (None, 0, 10, 10)],
```
Comment (NicolasHug) [inline test/test_metadata.py]: Nit for completeness, I think we can also test the case when only `begin_stream_seconds_from_content` is None?

```suggestion
    [(60, 5, 20, 15), (60, 1, None, 60), (60, None, 1, 60), (None, 0, 10, 10)],
```

File: test/test_metadata.py [hunk 6]
```diff
@@ -143,6 +143,93 @@ def test_num_frames_fallback(
...
+        average_fps_from_header=5,
+        stream_index=0,
+    )
+
+    assert metadata.duration_seconds == expected_duration_seconds
+
+
+@pytest.mark.parametrize(
+    "num_frames_from_header, average_fps_from_header, expected_duration_seconds",
+    [(100, 10, 10), (100, None, None), (None, None, None)],
```
Comment (NicolasHug) [inline test/test_metadata.py]: Similarly here we may want to test when only `num_frames_from_header` is missing:

```suggestion
    [(100, 10, 10), (100, None, None),  (None, 10, None), (None, None, None)],
```

File: src/torchcodec/_core/_metadata.py [hunk 7]
```diff
@@ -88,14 +88,22 @@ def duration_seconds(self) -> Optional[float]:
         fall back to ``duration_seconds_from_header``.
```
Comment (NicolasHug) [inline src/torchcodec/_core/_metadata.py]: Similarly to what was done for the `num_frames` properties, we can slightly update the docstring above to describe the new fall-back

**Review by NicolasHug — APPROVED** [review]
Thanks for the PR, and great work on the tests/mocking!

File: src/torchcodec/_core/_metadata.py [hunk 8]
```diff
@@ -85,17 +85,27 @@ class VideoStreamMetadata(StreamMetadata):
     def duration_seconds(self) -> Optional[float]:
         """Duration of the stream in seconds. We try to calculate the duration
         from the actual frames if a :term:`scan` was performed. Otherwise we
-        fall back to ``duration_seconds_from_header``.
+        fall back to ``duration_seconds_from_header``. If that value is None,
```
Comment (scotts) [inline src/torchcodec/_core/_metadata.py]: Nit:
```suggestion
        fall back to ``duration_seconds_from_header``. If that value is also None,
```

File: src/torchcodec/_core/_metadata.py [hunk 9]
```diff
@@ -123,14 +133,22 @@ def end_stream_seconds(self) -> Optional[float]:
 
     @property
     def num_frames(self) -> Optional[int]:
-        """Number of frames in the stream. This corresponds to
-        ``num_frames_from_content`` if a :term:`scan` was made, otherwise it
-        corresponds to ``num_frames_from_header``.
+        """Number of frames in the stream (int or None).
+        This corresponds to ``num_frames_from_content`` if a :term:`scan` was made,
+        otherwise it corresponds to ``num_frames_from_header``. If it is None,
```
Comment (scotts) [inline src/torchcodec/_core/_metadata.py]: Nit:
```suggestion
        otherwise it corresponds to ``num_frames_from_header``. If that value is also None,
```

**Review by scotts — APPROVED** [review]
All the new stuff looks good too, good to merge!
